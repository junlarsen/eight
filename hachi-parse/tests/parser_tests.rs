use hachi_parse::Lexer;
use hachi_parse::Parser;
use insta::assert_ron_snapshot;

#[test]
fn test_snapshot_corpus() {
    insta::glob!("data/*.test", |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let mut parser = Parser::new(Lexer::new(&input));
        let translation_unit = parser
            .parse()
            .unwrap_or_else(|_| panic!("failed to parse corpus file {} into ast", path.display()));
        assert_ron_snapshot!(*translation_unit);
    })
}
