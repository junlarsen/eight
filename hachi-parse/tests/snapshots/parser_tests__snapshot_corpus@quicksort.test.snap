---
source: hachi-parse/tests/parser_tests.rs
expression: "*translation_unit"
input_file: hachi-parse/tests/data/quicksort.test
snapshot_kind: text
---
TranslationUnit(
  ast_node_id: NodeId(0),
  items: [
    Function(FunctionItem(
      ast_node_id: NodeId(56),
      span: Span(
        low: 0,
        high: 152,
      ),
      name: Identifier(
        ast_node_id: NodeId(1),
        name: "quicksort",
        span: Span(
          low: 3,
          high: 12,
        ),
      ),
      parameters: [
        FunctionParameterItem(
          ast_node_id: NodeId(6),
          span: Span(
            low: 13,
            high: 21,
          ),
          name: Identifier(
            ast_node_id: NodeId(2),
            name: "xs",
            span: Span(
              low: 13,
              high: 15,
            ),
          ),
          type: Pointer(PointerType(
            ast_node_id: NodeId(5),
            span: Span(
              low: 17,
              high: 21,
            ),
            inner: Integer32(Integer32Type(
              ast_node_id: NodeId(4),
              span: Span(
                low: 18,
                high: 21,
              ),
            )),
          )),
        ),
        FunctionParameterItem(
          ast_node_id: NodeId(10),
          span: Span(
            low: 23,
            high: 29,
          ),
          name: Identifier(
            ast_node_id: NodeId(7),
            name: "l",
            span: Span(
              low: 23,
              high: 24,
            ),
          ),
          type: Integer32(Integer32Type(
            ast_node_id: NodeId(9),
            span: Span(
              low: 26,
              high: 29,
            ),
          )),
        ),
        FunctionParameterItem(
          ast_node_id: NodeId(14),
          span: Span(
            low: 31,
            high: 37,
          ),
          name: Identifier(
            ast_node_id: NodeId(11),
            name: "r",
            span: Span(
              low: 31,
              high: 32,
            ),
          ),
          type: Integer32(Integer32Type(
            ast_node_id: NodeId(13),
            span: Span(
              low: 34,
              high: 37,
            ),
          )),
        ),
      ],
      return_type: None,
      body: [],
    )),
    Function(FunctionItem(
      ast_node_id: NodeId(159),
      span: Span(
        low: 154,
        high: 394,
      ),
      name: Identifier(
        ast_node_id: NodeId(57),
        name: "partition",
        span: Span(
          low: 157,
          high: 166,
        ),
      ),
      parameters: [
        FunctionParameterItem(
          ast_node_id: NodeId(62),
          span: Span(
            low: 167,
            high: 175,
          ),
          name: Identifier(
            ast_node_id: NodeId(58),
            name: "xs",
            span: Span(
              low: 167,
              high: 169,
            ),
          ),
          type: Pointer(PointerType(
            ast_node_id: NodeId(61),
            span: Span(
              low: 171,
              high: 175,
            ),
            inner: Integer32(Integer32Type(
              ast_node_id: NodeId(60),
              span: Span(
                low: 172,
                high: 175,
              ),
            )),
          )),
        ),
        FunctionParameterItem(
          ast_node_id: NodeId(66),
          span: Span(
            low: 177,
            high: 183,
          ),
          name: Identifier(
            ast_node_id: NodeId(63),
            name: "l",
            span: Span(
              low: 177,
              high: 178,
            ),
          ),
          type: Integer32(Integer32Type(
            ast_node_id: NodeId(65),
            span: Span(
              low: 180,
              high: 183,
            ),
          )),
        ),
        FunctionParameterItem(
          ast_node_id: NodeId(70),
          span: Span(
            low: 185,
            high: 191,
          ),
          name: Identifier(
            ast_node_id: NodeId(67),
            name: "r",
            span: Span(
              low: 185,
              high: 186,
            ),
          ),
          type: Integer32(Integer32Type(
            ast_node_id: NodeId(69),
            span: Span(
              low: 188,
              high: 191,
            ),
          )),
        ),
      ],
      return_type: Some(Integer32(Integer32Type(
        ast_node_id: NodeId(72),
        span: Span(
          low: 195,
          high: 198,
        ),
      ))),
      body: [],
    )),
    Function(FunctionItem(
      ast_node_id: NodeId(191),
      span: Span(
        low: 396,
        high: 464,
      ),
      name: Identifier(
        ast_node_id: NodeId(160),
        name: "swap",
        span: Span(
          low: 399,
          high: 403,
        ),
      ),
      parameters: [
        FunctionParameterItem(
          ast_node_id: NodeId(165),
          span: Span(
            low: 404,
            high: 411,
          ),
          name: Identifier(
            ast_node_id: NodeId(161),
            name: "a",
            span: Span(
              low: 404,
              high: 405,
            ),
          ),
          type: Pointer(PointerType(
            ast_node_id: NodeId(164),
            span: Span(
              low: 407,
              high: 411,
            ),
            inner: Integer32(Integer32Type(
              ast_node_id: NodeId(163),
              span: Span(
                low: 408,
                high: 411,
              ),
            )),
          )),
        ),
        FunctionParameterItem(
          ast_node_id: NodeId(170),
          span: Span(
            low: 413,
            high: 420,
          ),
          name: Identifier(
            ast_node_id: NodeId(166),
            name: "b",
            span: Span(
              low: 413,
              high: 414,
            ),
          ),
          type: Pointer(PointerType(
            ast_node_id: NodeId(169),
            span: Span(
              low: 416,
              high: 420,
            ),
            inner: Integer32(Integer32Type(
              ast_node_id: NodeId(168),
              span: Span(
                low: 417,
                high: 420,
              ),
            )),
          )),
        ),
      ],
      return_type: None,
      body: [],
    )),
  ],
)
